src/main/java/com/societegenerale/commons/plugin/rules/DontReturnNullCollectionTest.java

```diff
@@ -0,0 +1,34 @@
+package com.societegenerale.commons.plugin.rules;
+
+import java.util.List;
+import java.util.Set;
+
+import javax.annotation.Nonnull;
+
+import com.societegenerale.commons.plugin.utils.ArchUtils;
+import com.tngtech.archunit.lang.ArchRule;
+
+import static com.tngtech.archunit.lang.syntax.ArchRuleDefinition.methods;
+
+/**
+ * Returning null collections (List, Set) forces the caller to always perform a null check, which hinders readability. It's much better to never return a null Collection, and instead return an empty one.
+ * This rule enforces that all methods returning a Collection must be annotated with @Nonnull
+ *
+ * @see: there is no agreed standard for notNull annotation, see <a href= "https://stackoverflow.com/questions/4963300/which-notnull-java-annotation-should-i-use/">here</a>
+ *
+ *
+ */
+public class DontReturnNullCollectionTest implements ArchRuleTest {
+
+  protected static final String NO_NULL_COLLECTION_MESSAGE = "we don't want callers to perform null check every time. Return an empty collection, not null.";
+
+  @Override
+  public void execute(String path) {
+
+    ArchRule rule = methods().that().haveRawReturnType(List.class).or().haveRawReturnType(Set.class).should().beAnnotatedWith(Nonnull.class)
```

Comment:
@codecholeric , do you think there's a better way to implement this rule ? I couldn't find a way to check for the use of Collection, so I am listing List and Set.. 
also, we may want to list other annotations than Nonnull, since there's no standard. how would you validate that a method is annotated with one of the accepted provided annotations ?

thanks !

Reply:
To replace List and Set, I think about :
`methods().that().haveRawReturnType(classes().that().implement(java.util.Collection.class))`

but I'll verify if it's correct

Reply:
note that following major refactoring, this should be implemented in https://github.com/societe-generale/arch-unit-build-plugin-core